### PR TITLE
[DAGCombine] Remove oneuse restrictions for RISCV in folding (shl (add_nsw x, c1)), c2) and folding (shl(sext(add x, c1)), c2) in some scenarios

### PR DESCRIPTION
**Pull request by LiqinWeng**
This patch remove the restriction for folding (shl (add_nsw x, c1)), c2)  and folding (shl(sext(add x, c1)), c2),  and test case from dhrystone , see this link:
riscv32: https://godbolt.org/z/o8GdMKrae
riscv64: https://godbolt.org/z/Yh5bPz56z

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp [hunk 1]
```diff
@@ -1036,6 +1036,11 @@ bool AMDGPUTargetLowering::isDesirableToCommuteWithShift(
   assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
           N->getOpcode() == ISD::SRL) &&
          "Expected shift op");
+
+  // if (!N->getOperand(0).hasOneUse()) {
+  //   return false;
+  // }
+
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:1043]: Commented out code

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 2]
```diff
@@ -10070,7 +10070,7 @@ SDValue DAGCombiner::visitSHL(SDNode *N) {
   // Variant of version done on multiply, except mul by a power of 2 is turned
   // into a shift.
   if ((N0.getOpcode() == ISD::ADD || N0.getOpcode() == ISD::OR) &&
-      N0->hasOneUse() && TLI.isDesirableToCommuteWithShift(N, Level)) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10073]: How is this beneficial if you have to duplicate the inner add?

File: llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp (hunk 1 above)
Comment (LiqinWeng) [reply llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp]: just for test ~~~have removed!

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 3]
```diff
@@ -10070,17 +10070,33 @@ SDValue DAGCombiner::visitSHL(SDNode *N) {
...
-      AddToWorklist(Shl0.getNode());
-      SDNodeFlags Flags;
-      // Preserve the disjoint flag for Or.
-      if (N0.getOpcode() == ISD::OR && N0->getFlags().hasDisjoint())
-        Flags.setDisjoint(true);
-      return DAG.getNode(N0.getOpcode(), DL, VT, Shl0, Shl1, Flags);
+      TLI.isDesirableToCommuteWithShift(N, Level)) {
+    // LD/ST will optimize constant Offset extraction, so when AddNode
+    // is used by LD/ST, it can still complete the folding optimization
+    // operation performed above.
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10076]: This depends on the legality and the specific offsets

File: llvm/test/CodeGen/RISCV/add_shl_constant.ll [hunk 4]
```diff
@@ -0,0 +1,60 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=riscv32 -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV32 %s
+; RUN: llc -mtriple=riscv64 -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV64 %s
+
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/add_shl_constant.ll:6]: Don't need -verify-machineinstrs

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 5]
```diff
@@ -10070,17 +10070,33 @@ SDValue DAGCombiner::visitSHL(SDNode *N) {
...
+      SDValue N01 = N0.getOperand(1);
+      if (SDValue Shl1 =
+              DAG.FoldConstantArithmetic(ISD::SHL, SDLoc(N1), VT, {N01, N1})) {
+        SDValue Shl0 =
+            DAG.getNode(ISD::SHL, SDLoc(N0), VT, N0.getOperand(0), N1);
+        AddToWorklist(Shl0.getNode());
+        SDNodeFlags Flags;
+        // Preserve the disjoint flag for Or.
+        if (N0.getOpcode() == ISD::OR && N0->getFlags().hasDisjoint())
+          Flags.setDisjoint(true);
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10097]: Just pass through the flags from N0->getFlags then?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 6]
```diff
@@ -10070,17 +10070,33 @@ SDValue DAGCombiner::visitSHL(SDNode *N) {
...
-      if (N0.getOpcode() == ISD::OR && N0->getFlags().hasDisjoint())
-        Flags.setDisjoint(true);
-      return DAG.getNode(N0.getOpcode(), DL, VT, Shl0, Shl1, Flags);
+      TLI.isDesirableToCommuteWithShift(N, Level)) {
+    // LD/ST will optimize constant Offset extraction, so when AddNode
+    // is used by LD/ST, it can still complete the folding optimization
+    // operation performed above.
+    bool canOptAwlays = false;
+    if (!N0.hasOneUse() && N0.getOpcode() == ISD::ADD) {
+      for (SDNode *Use : N0->uses()) {
```
[the remark this answers is not in the document]
Comment (LiqinWeng) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10079]: Thank you, I will try it.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 3 above)
Comment (LiqinWeng) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:10076]: image
If I must do this optimization, where do you think it would be better to do it?

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 7]
```diff
@@ -59541,3 +59541,18 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
...
+    const SDNode *N, CombineLevel Level) const {
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  if ((ShiftLHS.getOpcode() == ISD::SIGN_EXTEND &&
+       !(ShiftLHS->hasOneUse() && ShiftLHS.getOperand(0)->hasOneUse())) ||
+      !ShiftLHS->hasOneUse())
+    return false;
```
Comment (goldsteinn) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:59555]: Here (and everywhere else) think it would be much cleaner to just do:

```
match ShiftLHS->hasOneUse() || match(ShiftLHS, m_SExtLike(m_OneUse(m_Value(X)))
```

NB: The `m_SExtLike` needs to be added to `SDPatternMatch.h`, its just `ZEXT` + `NNEG`.

Comment (goldsteinn) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:59555]: Forget the `m_SextLike` (just do `m_SExt`) stuff, that will require more plumbing that in reasonable for this patch.

Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:59555]: image
I feel that this makes it even less clear. What do you think?

Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:59555]: why do you need all the llvm::SDPatternMatch:: namespace prefixes?

Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:59555]: image

Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:59555]: Have any other questions? Can I merge it？

Comment (RKSimon) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:59555]: why not just add a local `using namespace llvm::SDPatternMatch` - we already do this in X86ISelLowering.cpp in several places (we get namespace clashes if we do it across the entire file).

Comment (goldsteinn) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:59555]: Just use one using at the top scope of the function.

Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: done~

Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: @goldsteinn  @RKSimon  Are there any other issues with the current MR?

File: llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp [hunk 8]
```diff
@@ -1037,6 +1037,13 @@ bool AMDGPUTargetLowering::isDesirableToCommuteWithShift(
   assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
           N->getOpcode() == ISD::SRL) &&
          "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  if ((ShiftLHS.getOpcode() == ISD::SIGN_EXTEND &&
+       !(ShiftLHS->hasOneUse() && ShiftLHS.getOperand(0)->hasOneUse())) ||
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:1043]: demorgan negated condition expressions

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 9]
```diff
@@ -59668,3 +59668,19 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
...
+bool X86TargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  using namespace llvm::SDPatternMatch;
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  SDValue Add;
+  if (sd_match(ShiftLHS, m_OneUse(m_SExt(m_OneUse(m_Value(Add))))) ||
```
Comment (goldsteinn) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:59681]: The rest of the helpers should use the sd variant.

Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: fixed

File: llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp (hunk 8 above)
Comment (LiqinWeng) [reply llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp]: fixed

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 2 above)
Comment (LiqinWeng) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp]: pls see:
riscv32: https://godbolt.org/z/o8GdMKrae
riscv64: https://godbolt.org/z/Yh5bPz56z

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 9 above)
Comment (goldsteinn) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: This does not seem done...

File: llvm/lib/Target/Hexagon/HexagonISelLowering.cpp [hunk 10]
```diff
@@ -2156,6 +2157,23 @@ bool HexagonTargetLowering::hasBitTest(SDValue X, SDValue Y) const {
...
 
+bool HexagonTargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  using namespace llvm::SDPatternMatch;
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  SDValue Add;
```
Comment (mshockwave) [inline llvm/lib/Target/Hexagon/HexagonISelLowering.cpp:2168]: It this value used?

File: llvm/lib/Target/Hexagon/HexagonISelLowering.cpp [hunk 11]
```diff
@@ -2156,6 +2157,23 @@ bool HexagonTargetLowering::hasBitTest(SDValue X, SDValue Y) const {
...
+
+  if (ShiftLHS->hasOneUse() ||
+      sd_match(ShiftLHS, m_OneUse(m_SExt(m_OneUse(m_Value(Add))))))
```
Comment (mshockwave) [inline llvm/lib/Target/Hexagon/HexagonISelLowering.cpp:2171]: what are you trying to achieve here? I think either `ShiftLHS->hasOneUse()` or the sd_match call is redundant

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 12]
```diff
@@ -19137,3 +19138,20 @@ Value *PPCTargetLowering::emitMaskedAtomicCmpXchgIntrinsic(
...
+
+bool PPCTargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  using namespace llvm::SDPatternMatch;
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  SDValue Add;
```
Comment (mshockwave) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:19150]: ditto

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp [hunk 13]
```diff
@@ -19137,3 +19138,20 @@ Value *PPCTargetLowering::emitMaskedAtomicCmpXchgIntrinsic(
...
+
+  if (ShiftLHS->hasOneUse() ||
```
Comment (mshockwave) [inline llvm/lib/Target/PowerPC/PPCISelLowering.cpp:19152]: ditto

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 14]
```diff
@@ -60244,3 +60244,20 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
...
+    const SDNode *N, CombineLevel Level) const {
+  using namespace llvm::SDPatternMatch;
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  SDValue Add;
+
+  if (ShiftLHS->hasOneUse() ||
```
Comment (mshockwave) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:60258]: ditto

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 15]
```diff
@@ -60244,3 +60244,20 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
...
+
+bool X86TargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  using namespace llvm::SDPatternMatch;
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  SDValue Add;
```
Comment (mshockwave) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:60256]: ditto

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 16]
```diff
@@ -17923,6 +17931,7 @@ AArch64TargetLowering::isDesirableToCommuteWithShift(const SDNode *N,
       }
     }
   }
+
```
Comment (mshockwave) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17934]: please don't make unnecessary change

**Review by mshockwave — COMMENTED** [review]
Please don't force push unless necessary, because some of the comments might be lost

File: llvm/lib/Target/Hexagon/HexagonISelLowering.cpp (hunk 11 above)
Comment (LiqinWeng) [reply llvm/lib/Target/Hexagon/HexagonISelLowering.cpp:2171]: This patch remove the restriction for folding (shl (add_nsw x, c1)), c2) and folding (shl(sext(add x, c1)), c2), and test case from dhrystone , see this link:
riscv32: https://godbolt.org/z/o8GdMKrae
riscv64: https://godbolt.org/z/Yh5bPz56z

**Comment by LiqinWeng** [comment]
> Please don't force push unless necessary, because some of the comments might be lost

Sorry, but I didn't change any commits. The local branch is updated to the latest version. Pushing it will give an error, so I forced to push.

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 9 above)
Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: Some architectures, such as ARM, use this logic（sd_match(ShiftLHS, m_OneUse(m_SExt(m_OneUse(m_Value(Add)） to directly return true, which doesn't seem appropriate, so I didn't change it.

File: llvm/lib/Target/Hexagon/HexagonISelLowering.cpp (hunk 11 above)
Comment (mshockwave) [reply llvm/lib/Target/Hexagon/HexagonISelLowering.cpp:2171]: What I meant was that if `ShiftLHS->hasOneUse()` is false, the following sd_match will also be false -- then why do you have both `ShiftLHS->hasOneUse()` _and_ the sd_match call?

**Comment by mshockwave** [comment]
> The local branch is updated to the latest version

Right, please avoid rebasing the local branch unless necessary, per LLVM's code review policy: https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

File: llvm/test/CodeGen/RISCV/add_shl_constant.ll (hunk 4 above)
Comment (LiqinWeng) [reply llvm/test/CodeGen/RISCV/add_shl_constant.ll]: fixed

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 16 above)
Comment (LiqinWeng) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17934]: removed

**Comment by LiqinWeng** [comment]
> > The local branch is updated to the latest version
> 
> Right, please avoid rebasing the local branch unless necessary, per LLVM's code review policy: https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

thanks

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 14 above)
Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:60258]: done

File: llvm/lib/Target/X86/X86ISelLowering.cpp (hunk 15 above)
Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:60256]: done

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 13 above)
Comment (LiqinWeng) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp]: done

File: llvm/lib/Target/PowerPC/PPCISelLowering.cpp (hunk 12 above)
Comment (LiqinWeng) [reply llvm/lib/Target/PowerPC/PPCISelLowering.cpp]: done

File: llvm/lib/Target/Hexagon/HexagonISelLowering.cpp (hunk 11 above)
Comment (LiqinWeng) [reply llvm/lib/Target/Hexagon/HexagonISelLowering.cpp]: fixed

File: llvm/lib/Target/Hexagon/HexagonISelLowering.cpp (hunk 10 above)
Comment (LiqinWeng) [reply llvm/lib/Target/Hexagon/HexagonISelLowering.cpp]: removed

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 17]
```diff
@@ -60468,3 +60468,20 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
...
+    const SDNode *N, CombineLevel Level) const {
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  if (!ShiftLHS->hasOneUse())
+    return false;
+
+  if ((ShiftLHS.getOpcode() == ISD::SIGN_EXTEND &&
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:60482]: remove extra ( )

Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp:60482]: fixed

**Comment by topperc** [comment]
For https://godbolt.org/z/o8GdMKrae, the gcc code for `add_shl_moreoneuse_store` looks like a direct translation of the original code. Is something in the middle end changing the IR and making it produce worse code?

**Comment by topperc** [comment]
> > For https://godbolt.org/z/o8GdMKrae, the gcc code for `add_shl_moreoneuse_store` looks like a direct translation of the original code. Is something in the middle end changing the IR and making it produce worse code?
> 
> I have tried doing it in the instcombine phase before, and there is a trend of performance degradation for some test cases on some architectures~Could you help me take a look at this issue：#112830

#112830 is for sjeng. I thought this patch was for dhrystone?

**Comment by LiqinWeng** [comment]
> #112830

Yes, this patch was for dhrystone~ Which sjeng's issue is for you to help see if you have any ideas~

**Comment by LiqinWeng** [comment]
> > > For https://godbolt.org/z/o8GdMKrae, the gcc code for `add_shl_moreoneuse_store` looks like a direct translation of the original code. Is something in the middle end changing the IR and making it produce worse code?
> > 
> > 
> > I have tried doing it in the instcombine phase before, and there is a trend of performance degradation for some test cases on some architectures~Could you help me take a look at this issue：#112830
> 
> #112830 is for sjeng. I thought this patch was for dhrystone?

Can this patch be merged?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 18]
```diff
@@ -18041,8 +18041,46 @@ bool RISCVTargetLowering::isDesirableToCommuteWithShift(
   //   (shl (or x, c1), c2) -> (or (shl x, c2), c1 << c2)
   SDValue N0 = N->getOperand(0);
   EVT Ty = N0.getValueType();
+
+  // LD/ST will optimize constant Offset extraction, so when AddNode is used by
+  // LD/ST, it can still complete the folding optimization operation performed
+  // above.
+  auto isLDST = [&]() {
+    bool canOptAwlays = false;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:18049]: Always*

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:18049]: Variable names should be capitalized

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 19]
```diff
@@ -18041,8 +18041,46 @@ bool RISCVTargetLowering::isDesirableToCommuteWithShift(
...
+        }
+        canOptAwlays = true;
+      }
+    }
+    return canOptAwlays;
+  };
+
   if (Ty.isScalarInteger() &&
       (N0.getOpcode() == ISD::ADD || N0.getOpcode() == ISD::OR)) {
+    if (N0.getOpcode() == ISD::ADD && !N0->hasOneUse()) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:18081]: Drop curly braces

Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: done

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 18 above)
Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: done

**Review by mshockwave — COMMENTED** [review]
Why do you override PowerPC and Hexagon's isDesirableToCommuteWithShift while they weren't previously overridden? Does the new behavior break their tests? If that's the case what about other targets?

**Comment by LiqinWeng** [comment]
> Why do you override PowerPC and Hexagon's isDesirableToCommuteWithShift while they weren't previously overridden? Does the new behavior break their tests? If that's the case what about other targets?

I actually didn't change the logic. I just moved the constraints to an interface of isDesirableToCommuteWithShift to implement them.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 20]
```diff
@@ -17910,6 +17910,13 @@ AArch64TargetLowering::isDesirableToCommuteWithShift(const SDNode *N,
   SDValue ShiftLHS = N->getOperand(0);
   EVT VT = N->getValueType(0);
 
+  if (!ShiftLHS->hasOneUse())
```
Comment (topperc) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17913]: Does this logic need to be added to the default implementation of `isDesirableToCommuteWithShift` for targets that don't override it? I believe the default returns `true`

Comment (LiqinWeng) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17913]: Some test cases will fail~Brief look, it seems that there is no benefit
image

Comment (topperc) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:17913]: AArch64 overrrides the hook. I wasn't asking you to change what you already did for AArch64. So why do AArch64 tests change?

I was asking you to additionally change the default implementation.

Comment (LiqinWeng) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: Any problems with this Patch？

Comment (topperc) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: There was a holiday in the US on Thursday so I'm a little behind on reviews right now.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 21]
```diff
@@ -18151,8 +18151,46 @@ bool RISCVTargetLowering::isDesirableToCommuteWithShift(
   //   (shl (or x, c1), c2) -> (or (shl x, c2), c1 << c2)
   SDValue N0 = N->getOperand(0);
   EVT Ty = N0.getValueType();
+
+  // LD/ST will optimize constant Offset extraction, so when AddNode is used by
+  // LD/ST, it can still complete the folding optimization operation performed
+  // above.
+  auto isLDST = [&]() {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:18158]: This function name isn't great. Should it be `isUsedByLdSt`?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 22]
```diff
@@ -18151,8 +18151,46 @@ bool RISCVTargetLowering::isDesirableToCommuteWithShift(
...
+    bool CanOptAlways = false;
+    if (N0->getOpcode() == ISD::ADD && !N0->hasOneUse()) {
+      for (SDNode *Use : N0->uses()) {
+        // This use is the one we're on right now. Skip it
+        if (Use == N || Use->getOpcode() == ISD::SELECT)
+          continue;
+        if (!isa<StoreSDNode>(Use) && !isa<LoadSDNode>(Use)) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:18165]: Do we need to check that `c1 << c2` fits in 12 bits so it will fold into the load/store?

Comment (LiqinWeng) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:18165]: There may be no need for check. The address calculation will generate LUI + addi to ensure that the LD/ST offset falls within the 12-bit range.

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 23]
```diff
@@ -60673,3 +60673,20 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
     return Align(1ULL << ExperimentalPrefInnermostLoopAlignment);
   return TargetLowering::getPrefLoopAlignment();
 }
+
+bool X86TargetLowering::isDesirableToCommuteWithShift(
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:60677]: remove this? it looks like it matched the generic implementation

Comment (LiqinWeng) [reply llvm/lib/Target/X86/X86ISelLowering.cpp]: done

**Comment by LiqinWeng** [comment]
Any other Porblems? @topperc @RKSimon

**Review by RKSimon — APPROVED** [review]
LGTM

**Review by topperc — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
Regression:
```
; bin/llc -mtriple=riscv64 -mattr=+zba test.ll -o -
define i64 @test(i64 %x) {
entry:
  %or = or i64 %x, 1
  %mul = mul nuw nsw i64 %or, 3
  ret i64 %mul
}
```
Before: https://godbolt.org/z/6vKhhxrWr
```
test:
        ori     a0, a0, 1
        sh1add  a0, a0, a0
        ret
```
After:
```
test:
        ori     a1, a0, 1
        slli    a0, a0, 1
        ori     a0, a0, 2
        add     a0, a0, a1
        ret
```

**Comment by lukel97** [comment]
> Regression:
> 
> ```
> 
> ; bin/llc -mtriple=riscv64 -mattr=+zba test.ll -o -
> 
> define i64 @test(i64 %x) {
> 
> entry:
> 
>   %or = or i64 %x, 1
> 
>   %mul = mul nuw nsw i64 %or, 3
> 
>   ret i64 %mul
> 
> }
> 
> ```
> 
> Before: https://godbolt.org/z/6vKhhxrWr
> 
> ```
> 
> test:
> 
>         ori     a0, a0, 1
> 
>         sh1add  a0, a0, a0
> 
>         ret
> 
> ```
> 
> After:
> 
> ```
> 
> test:
> 
>         ori     a1, a0, 1
> 
>         slli    a0, a0, 1
> 
>         ori     a0, a0, 2
> 
>         add     a0, a0, a1
> 
>         ret
> 
> ```


Has there been a fix for this yet? If not it would be good to create an issue so we don't forget about it.

**Comment by dtcxzyw** [comment]
No. Thank you for doing this!

**Comment by lukel97** [comment]
Opened #119527 to address this
